### PR TITLE
Add Settings Page and bring GraphiQL IDE and WPGraphQL Insights into the core plugin

### PR DESCRIPTION
**Pull request by jasonbahl**
This PR does the following: 

- Adds GraphiQL IDE to the main WPGraphQL plugin (deprecates the need for separate [WPGraphiQL](https://github.com/wp-graphql/wp-graphiql) plugin)
- Adds GraphQL Settings Page with API to register more fields and field sections
- Adds Query Logs and Tracing to the main WPGraphQL plugin (deprecates the need for separate [WPGraphQL Insights](https://github.com/wp-graphql/wp-graphql-insights) plugin)

## GraphiQL IDE

WPGraphQL now has GraphiQL IDE built in to WPGraphQL! 

GraphiQL IDE is an incredibly useful tool for working with GraphQL APIs, and now it's available to all users of WPGraphQL without needing to find and install a separate plugin. 

With WPGraphQL Active, there is now an Admin Menu for GraphQL, and within that menu is "GraphiQL IDE" link

Screen Shot 2020-09-16 at 3 10 44 PM

There's also an Admin Bar short-link for GraphiQL IDE: 

Screen Shot 2020-09-16 at 3 11 42 PM

Clicking either link will open GraphiQL in the WP Admin: 

Screen Shot 2020-09-16 at 3 12 19 PM

You can now browse the WPGraphQL Schema and test Queries and Mutations all from within the WordPress Dashboard. 

## Settings Page (and API)
WPGraphQL now has a settings page where some common configurations can be set, without needing to touch code. 

In the Admin Menu, there's now a "GraphQL > Settings" link

Screen Shot 2020-09-16 at 3 17 20 PM

Clicking that will take you to a Settings Page: 

Screen Shot 2020-09-16 at 3 18 05 PM

### Settings Overview

These settings can be used to accomplish some common tasks with WPGraphQL that previously required code (filters/constants) to accomplish. 

- **GraphQL Endpoint**: You can now change the endpoint to be something other than `/graphql`.
- **Enable GraphiQL IDE**: You can enable/disable GraphiQL IDE in the Admin. Now that GraphiQL is bundled, you can turn it off if you prefer not to have it. 
- **Delete Data on Deactivation**: You can opt in or out of having any data WPGraphQL stores be deleted when WPGraphQL is de-activated. Things like the settings for this settings page get stored in the database. You can choose to keep data (for example temporary de-activations) or delete all data on de-activation to keep your data clean.
- **Enable GraphQL Debug Mode**: When errors occur in GraphQL requests, the default message is "Internal Server Error" to ensure implementation details are not leaked to the public. Enabling GraphQL Debug Mode shows more detailed messages, allowing developers to identify issues more quickly. 
- **Enable GraphQL Tracing**: Enabling tracing will track performance of the entire request and each field within the request. The trace data will be returned as part of the "extensions" portion of the GraphQL response, alongside the data. This feature was previously part of WPGraphQL Insights, but can now be enabled by checking this setting. Trace data can be helpful for identifying bottlenecks (slow resolvers)
- **Tracing Role:** With Tracing Enabled, it can be limited to a certain roles. Only the selected role (or all) will see the trace data in results of GraphQL requests
- **Enable Query Logs**: Enabling Query Logs will return MySQL Query logs in the "extensions" portion of the GraphQL response. This is essentially like Query Monitor for WPGraphQL. You can see the specific MySQL queries being executed to resolve GraphQL requests.
- **Query Log Role**: If Query Logs are enabled, they will execute, but the results can be limited to a specific user role, or all user roles (public requests allowed). 

### Settings API

Along with the Settings Page, there are some basic APIs to allow 3rd party plugins to easily add GraphQL Settings. 

#### register_graphql_settings_section

This function should be used within the `graphql_register_settings` action, and allows new Settings Sections to be registered to the Settings Page. 

**usage:**

```php
add_action( 'graphql_register_settings', function() {
	register_graphql_settings_section( 'graphql_my_section', [
		'title' => __( 'My Section', 'your-textdomain' ),
	]);
});
```

This will add a new section to the settings page like so: 

Screen Shot 2020-09-16 at 3 40 56 PM

#### register_graphql_settings_field

This function will register a new settings field to the specified settings section: 

**usage:**

```php
add_action( 'graphql_register_settings', function() {
	register_graphql_settings_field( 'graphql_my_section', [
		'name'    => 'my_field',
		'label'   => __( 'My Field', 'your-textdomain' ),
		'desc'    => __( 'Your field description...', 'your-textdomain' ),
		'type'    => 'text',
		'default' => "my default value...",
	]);
});
```

This registers a text field named with the "My Field" label to the "graphql_my_section" section. 

Screen Shot 2020-09-16 at 3 44 37 PM

#### register_graphql_settings_fields

This allows multiple settings fields to be registered to a section. 

**usage:**

```php
add_action( 'graphql_register_settings', function() {
	register_graphql_settings_section( 'graphql_new_section', [
		'title' => __( 'New Section', 'your-textdomain' ),
	] );
	register_graphql_settings_fields( 'graphql_new_section', [
		[
			'name'    => 'text',
			'label'   => __( 'Text Input', 'wedevs' ),
			'desc'    => __( 'Text input description', 'wedevs' ),
			'type'    => 'text',
			'default' => 'Title'
		],
		[
			'name'  => 'textarea',
			'label' => __( 'Textarea Input', 'wedevs' ),
			'desc'  => __( 'Textarea description', 'wedevs' ),
			'type'  => 'textarea'
		],
		[
			'name'  => 'checkbox',
			'label' => __( 'Checkbox', 'wedevs' ),
			'desc'  => __( 'Checkbox Label', 'wedevs' ),
			'type'  => 'checkbox'
		],
		[
			'name'    => 'radio',
			'label'   => __( 'Radio Button', 'wedevs' ),
			'desc'    => __( 'A radio button', 'wedevs' ),
			'type'    => 'radio',
			'options' => [
				'yes' => 'Yes',
				'no'  => 'No'
			]
		],
		[
			'name'    => 'multicheck',
			'label'   => __( 'Multile checkbox', 'wedevs' ),
			'desc'    => __( 'Multi checkbox description', 'wedevs' ),
			'type'    => 'multicheck',
			'options' => [
				'one'   => 'One',
				'two'   => 'Two',
				'three' => 'Three',
				'four'  => 'Four'
			]
		],
		[
			'name'    => 'selectbox',
			'label'   => __( 'A Dropdown', 'wedevs' ),
			'desc'    => __( 'Dropdown description', 'wedevs' ),
			'type'    => 'select',
			'default' => 'no',
			'options' => [
				'yes' => 'Yes',
				'no'  => 'No'
			]
		],
		[
			'name'    => 'password',
			'label'   => __( 'Password', 'wedevs' ),
			'desc'    => __( 'Password description', 'wedevs' ),
			'type'    => 'password',
			'default' => ''
		],
		[
			'name'    => 'file',
			'label'   => __( 'File', 'wedevs' ),
			'desc'    => __( 'File description', 'wedevs' ),
			'type'    => 'file',
			'default' => ''
		],
		[
			'name'    => 'color',
			'label'   => __( 'Color', 'wedevs' ),
			'desc'    => __( 'Color description', 'wedevs' ),
			'type'    => 'color',
			'default' => ''
		],
		[
			'name'    => 'user_role',
			'label'   => __( 'User Role', 'wp-graphql' ),
			'desc'    => __( 'Description for selecting a User Role.', 'wp-graphql' ),
			'type'    => 'user_role_select',
			'default' => "administrator",
		],
	] );
} );
```

Screen Shot 2020-09-16 at 3 49 51 PM

#### get_graphql_setting

This function should be used to get settings values from GraphQL Settings pages. Each settings section is stored in a serialized array in the options table. This function will properly get the value and return it. 

**usage:**

If you wanted to get the "color" setting from the "graphql_new_section" section, and have it default to "blue" if no value is set, you would do so like: 

```php
$option_name = 'color';
$default_value = 'blue';
$section = 'graphql_new_section';
$value = get_graphql_setting( $option_name, $default_value, $section );
```

## Tracing

We saw that there are settings for Tracing, but what exactly is Tracing? 

Tracing is performance tracking for the entire GraphQL request _and_ each field of the request. This data can be helpful when debugging bottlenecks. 

With Tracing enabled from the Settings Page, executing a GraphQL Query will show the trace data in the "extensions" portion of the request: 

Screen Shot 2020-09-16 at 3 56 31 PM

There is data for the entire request: 

- startTime
- endTime
- duration

And data for each field within the request:

- path
- parentType
- fieldName
- returnType
- startOffset
- duration

This data can be very helpful when debugging bottlenecks. 

Let's take a look at an example. 

Executing the following query: 

```graphql
{
  posts {
    nodes {
      id
      title
    }
  }
}
```

We would see the following Trace data for the request: 

```json
  ...
  "tracing": {
  "version": 1,
  "startTime": "1600294443.335",
  "endTime": "1600294443.381",
  "duration": 45998,
  "execution": {
      ...
```

The request duration is `45998` microseconds (`0.045998` seconds).

And if we scrolled down, we'd see the following trace data for the `id` field of the first post:

```json
...,
{
  "path": [
    "posts",
    "nodes",
    0,
    "title"
  ],
  "parentType": "Post",
  "fieldName": "title",
  "returnType": "String",
  "startOffset": 27518,
  "duration": 797
},
...
```

Here, we can see that the duration was `797` microseconds (`0.000797 seconds`)

Let's override this field resolver like so: 

```php
add_filter( 'graphql_Post_fields', function( $fields ) {

	$fields['title']['resolve'] = function( \WPGraphQL\Model\Post $post ) {

		// Sleep for 1 second to simulate something taking a long time
		sleep(1);
		return $post->titleRendered;
	};

	return $fields;

});
```

Now, if we look at the trace data, we can see: 

```json
"tracing": {
 "version": 1,
 "startTime": "1600294618.1841",
 "endTime": "1600294628.2435",
 "duration": 10059434,
```
The entire request duration is now `10059434` microseconds ( `10.059434` seconds). 

Something is clearly wrong! This request used to take well under 1 second, and now it's taking more than 10 seconds! 😱

If we scroll down we will see that the duration of each title field is looking quite long:

```json
```json
...
{
  "path": [
    "posts",
    "nodes",
    0,
    "title"
  ],
  "parentType": "Post",
  "fieldName": "title",
  "returnType": "String",
  "startOffset": 27518,
  "duration": 1000964
},
...
```

The duration for the title field is now `1000964` microseconds (`1.000964` seconds). 

Now, we as developers know what field to start troubleshooting. 

In our case we know that the title field had a `sleep(1)` in the resolver, so we can remove that and we're back to super speedy execution! 

## Query Logs
Query Logs are a super helpful debugging tool. A lot of things happen under the hood to resolve a GraphQL Query, and sometimes it can be a mystery. 

Enabling Query Logs can give insight into what SQL Queries are being executed, how long they're taking, the raw SQL that is executed and the stacktrace to where in the codebase the query is being executed from. 

This can be super helpful when things don't seem to be behaving as you'd expect. Sometimes you get results that you don't expect, and analyzing the raw SQL queries can help identify _why_ things are misbehaving. Sometimes it's a rogue plugin that's filtering WP_Query args too aggressively, or perhaps a GraphQL input argument isn't properly mapping to WP_Query. 

In any case, seeing the raw SQL queries and the path to the code that called it can be a huge time saver when debugging. 

Screen Shot 2020-09-16 at 4 26 35 PM

### CODE REVIEW

**Comment by TylerBarnes** [comment]
Awesome! :D

File: src/Utils/QueryLog.php [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+	 * @return array
+	 */
+	public function show_results( $response, $schema, $operation_name, $request, $variables ) {
+
+		$query_log = $this->get_query_log();
+
+		// If the user cannot see the logs, return the response as-is without the logs
+//		if ( ! $this->user_can_see_logs() ) {
+//			return $response;
+//		}
```
Comment (renatonascalves) [inline src/Utils/QueryLog.php:109]: Is this comment is still valid?

File: src/Admin/Settings/SettingsRegistry.php [hunk 2]
```diff
@@ -0,0 +1,707 @@
...
+	 */
+	function register_fields( $section, $fields ) {
+		foreach ( $fields as $field ) {
+			$this->register_field( $section, $field );
+		}
+
+		return $this;
+	}
+
+	function register_field( $section, $field ) {
```
Comment (renatonascalves) [inline src/Admin/Settings/SettingsRegistry.php:89]: Missing phpdoc.

File: src/Admin/Settings/SettingsRegistry.php [hunk 3]
```diff
@@ -0,0 +1,707 @@
...
+	}
+
+	/**
+	 * Get field description for display
+	 *
+	 * @param array $args settings field args
+	 *
+	 * @return string
+	 */
+	public function get_field_description( $args ) {
```
Comment (renatonascalves) [inline src/Admin/Settings/SettingsRegistry.php:184]:
```suggestion
	public function get_field_description( $args ): string {
```

File: src/Admin/Settings/Settings.php [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+		add_submenu_page(
+			'graphql',
+			__( 'WPGraphQL Options', 'wp-graphql' ),
+			__( 'Settings', 'wp-graphql' ),
+			'manage_options',
+			'graphql',
+			[ $this, 'render_settings_page' ]
+		);
+
+
```
Comment (renatonascalves) [inline src/Admin/Settings/Settings.php:51]: A bit of spacing here....

File: src/Utils/Tracing.php [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+		return apply_filters( 'graphql_user_can_see_trace_data', $can_see );
+
+	}
+
+	/**
+	 * Get the trace to add to the response
+	 *
+	 * @return array
+	 */
+	public function get_trace() {
```
Comment (renatonascalves) [inline src/Utils/Tracing.php:329]:
```suggestion
	public function get_trace(): array {
```

File: src/Utils/Tracing.php [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+			]
+		];
+
+		/**
+		 * Filter the trace
+		 *
+		 * @param array   $trace The trace to return
+		 * @param Tracing $this  The Tracing class instance
+		 */
+		return apply_filters( 'graphql_tracing_response', $trace, $this );
```
Comment (renatonascalves) [inline src/Utils/Tracing.php:348]:
```suggestion
		return apply_filters( 'graphql_tracing_response', (array) $trace, $this );
```

File: src/Utils/Tracing.php [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+	}
+
+	/**
+	 * Determine if the requesting user can see trace data
+	 *
+	 * @return boolean
+	 */
+	public function user_can_see_trace_data() {
+
+		return true;
```
Comment (renatonascalves) [inline src/Utils/Tracing.php:296]: Leftovers?

```suggestion
```

File: src/Utils/Tracing.php [hunk 8]
```diff
@@ -0,0 +1,350 @@
...
+	public function get_request_duration() {
+		return ( $this->request_end_microtime - $this->request_start_microtime ) * 1000000;
+	}
+
+	/**
+	 * Determine if the requesting user can see trace data
+	 *
+	 * @return boolean
+	 */
+	public function user_can_see_trace_data() {
```
Comment (renatonascalves) [inline src/Utils/Tracing.php:294]:
```suggestion
	public function user_can_see_trace_data(): bool {
```

File: src/Utils/Tracing.php [hunk 9]
```diff
@@ -0,0 +1,350 @@
...
+		return $timestamp->format( "Y-m-d\TH:i:s.uP" );
+	}
+
+	/**
+	 * Filter the headers that WPGraphQL returns to include headers that indicate the WPGraphQL
+	 * server supports Apollo Tracing and Credentials
+	 *
+	 * @param $headers
+	 *
+	 * @return mixed
```
Comment (renatonascalves) [inline src/Utils/Tracing.php:237]:
```suggestion
	 * @return array
```

File: src/Utils/Tracing.php [hunk 10]
```diff
@@ -0,0 +1,350 @@
...
+	 */
+	public function return_tracing_headers( $headers ) {
```
Comment (renatonascalves) [inline src/Utils/Tracing.php:239]:
```suggestion
	public function return_tracing_headers( $headers ): array {
```

File: src/Utils/Tracing.php [hunk 11]
```diff
@@ -0,0 +1,350 @@
...
+		$headers[] = 'X-Insights-Include-Tracing';
+		$headers[] = 'X-Apollo-Tracing';
+		$headers[] = 'Credentials';
+
+		return $headers;
```
Comment (renatonascalves) [inline src/Utils/Tracing.php:244]:
```suggestion
		return (array) $headers;
```

File: src/Utils/Tracing.php [hunk 12]
```diff
@@ -0,0 +1,350 @@
...
+
+	/**
+	 * The version of the Apollo Tracing Spec
+	 * @var int
+	 */
+	public $trace_spec_version = 1;
+
+	/**
+	 * The user role tracing is limited to
+	 * @var
```
Comment (renatonascalves) [inline src/Utils/Tracing.php:68]: What's this variable type?

File: src/Utils/Tracing.php [hunk 13]
```diff
@@ -0,0 +1,350 @@
...
+
+	/**
+	 * The end timestamp
+	 * @var
+	 */
+	public $request_end_timestamp;
+
+	/**
+	 * The trace for the current field being resolved
+	 * @var
```
Comment (renatonascalves) [inline src/Utils/Tracing.php:56]:
```suggestion
	 * @var array
```

**Review by renatonascalves — CHANGES_REQUESTED** [review]
A few suggestions!

I also noted lots of opportunities to improve type casting, PHPdocs improvements, and code standardization here.

**Pull request merged**